shared/utils/typedefs-string.hpp

```diff
@@ -102,19 +117,110 @@ struct StringW {
     constexpr StringW(std::nullptr_t npt) noexcept : inst(npt) {}
     constexpr StringW() noexcept : inst(nullptr) {}
 
-    operator std::string();
+    constexpr void const* convert() const noexcept {
+        return inst;
+    }
     constexpr void* convert() noexcept {
         return inst;
     }
+    constexpr operator Il2CppString const*() const noexcept {
+        return inst;
+    }
     constexpr operator Il2CppString*() noexcept {
         return inst;
     }
+    constexpr Il2CppString const* operator->() const noexcept {
+        return inst;
+    }
     constexpr Il2CppString* operator->() noexcept {
         return inst;
     }
-    operator std::u16string();
-    operator std::wstring();
+    constexpr operator bool() const noexcept {
+        return inst != nullptr; 
+    }
+
+    // the argument here is unused but we don't care. the comparison against nullptr more clearly shows what this operator is for imo
+    #pragma clang diagnostic push
+    #pragma clang diagnostic ignored "-Wunused-parameter"
+    constexpr bool operator ==(std::nullptr_t rhs) const noexcept {
+        return inst == nullptr;
+    }
+    #pragma clang diagnostic pop
+
+    template<typename T>
+    requires (!std::is_convertible_v<T, Il2CppString*> && (std::is_constructible_v<std::u16string_view, T> || std::is_constructible_v<std::string_view, T>))
```

Comment:
Might make more sense for this constraint to essentially say: "can I construct a StringW from this" instead

Reply:
Adding on this, it should be possible to just say `inst = strappend(inst, rhs)` etc. instead of making a StringW and then casting back to an Il2CppString*